services/wallet/routeexecution/manager.go

```diff
@@ -85,6 +85,10 @@ func (m *Manager) BuildTransactionsFromRoute(ctx context.Context, buildInputPara
 			err = ErrCannotResolveRouteId
 			return
 		}
+		if m.buildInputParams != nil && m.buildInputParams.Uuid != buildInputParams.Uuid {
```

Comment:
@dlipicar I see, this fixes the issue, but in my opinion, it should be done earlier.

For example:
- the user runs swap flow with the entered amount `10 USDC` and `GetSuggestedRoutesAsync` is called
- the user wants to proceed and confirms Approval for `10 USDC` 
- now the user needs to confirm Swap tx, but changed his mind and wants to swap `15 USDC`, at this moment we call `GetSuggestedRoutesAsync` and that's the moment when we need to clear local data from this manager.

So what I would suggest as a change is to remove this code from here and add this line `api.s.routeExecutionManager.ClearLocalRouteData()` to `GetSuggestedRoutesAsync` function, that it looks like this:
```go
func (api *API) GetSuggestedRoutesAsync(ctx context.Context, input *requests.RouteInputParams) {
	logutils.ZapLogger().Debug("call to GetSuggestedRoutesAsync")

	api.s.routeExecutionManager.ClearLocalRouteData()

	api.s.router.SuggestedRoutesAsync(input)
}
```

Reply:
hmm I'm in general against adding logic to `api.go`.

Also I believe we should move in the direction of decoupling the route generation and route execution flows as much as possible. That is, is there a good reason to -by design- prevent a client from resuming a previously started route execution when the generation of some other one is requested? Not saying it's possible at the moment anyway, but it could be if the route executor stored the route it's in the process of executing instead of getting whatever the router generated the last time.

Reply:
We need to pass much more data from the cline at each step if we would decouple them.

Also statusgo logic is an unbreakable part of the app, so the client (desktop or mobile) works together with statusgo as a whole. Temporarily storing data in local memory should not be an issue.

Suppose we want to restore previously started route execution. In that case, we need to extend our statusgo logic and keep data not just locally but in db, and remind the user later and/or after the app restart that he has some unfinished flow. But that's another topic.

I agree that we should not keep logic in `api.go`, but with adding this line `api.s.routeExecutionManager.ClearLocalRouteData()` the function `GetSuggestedRoutesAsync` is still pretty clean. So all we did there is that we cleared any previously run flow (cause step 2 below relies on the data generated in the step 1, step 3 on step 2, step 4 on step 3). 
Our send flow is composed of the following steps:
1. ask for the best route
2. build the tx/s from the best route
3. sign tx/s
4. send tx/s

Without having this call `api.s.routeExecutionManager.ClearLocalRouteData()` in `GetSuggestedRoutesAsync` data generated from the previous step will remain in the local memory until the next run of the Send modal or till the app's lifetime.

Also we should call `api.s.routeExecutionManager.ClearLocalRouteData()` from `StopSuggestedRoutesAsyncCalculation`.

Reply:
@saledjenic 

First, I understand that, the way things currently are, the route execution process will be lost if for any reason the client requests a new suggested route, both with my and your proposed change.

`We need to pass much more data from the client at each step if we would decouple them.`

I'm not proposing that or any API breaking change at the moment. 
In general terms, I think smaller, more decoupled modules = better, and I think we could work towards making the two modules (generation and execution) more decoupled if, once the generator's job is done, the executor were in charge of holding the route (the data still lives in status-go). AFAIU we don't expect anything else from the router once the tx generation is done, so we can fully stop depending on it.

The thing about being able to resume a partial route execution regardless of whether the client requested some route is also not something I'm proposing we do now, just trying to give an example of the executor "owning" the route making more sense. Persisting stuff in the DB would be a requirement if you expect that process to work over an app restart, but that's orthogonal to the idea we're discussing.

I get that execution (steps 2,3 and 4) needs to happen after generation (step 1), I don't see why the executor needs to keep depending on the generator after it gets the data it needs from it.

-o-

Anyway, I feel this is more than enough discussion for a 4 line PR. I'm not particularly worried about whether the function call is in one place or the other, it doesn't seem to make any practical difference, so I'll just do the change you suggested. I am a bit concerned about the lack of test coverage in this part of the code and want us to make design decisions that help those tests be easier to write.

Reply:
>  more decoupled modules = better

I agree.

> once the generator's job is done, the executor were in charge of holding the route

I agree, but there is no use case now when we need this. If we want to let user resume already stated sending then we need additional logic for that which can be developed when we decide to go that way.

> AFAIU we don't expect anything else from the router once the tx generation is done, so we can fully stop depending on it.

I agree, but those data live in memory until the next Send flow run or until the app closes. There is no use case for that for now.

> Anyway, I feel this is more than enough discussion for a 4 line PR.

I agree.

Regarding tests, hope to have time soon to start adding them and removing those "TestsMode" checks.

Finally, shouldn't we have the same call to clear data from `StopSuggestedRoutesAsyncCalculation`?

Reply:
@saledjenic as far as I can see, `StopSuggestedRoutesAsyncCalculation` just stops the auto fees refresh, but it doesn't clear the last calculated route so executing it will still work fine.

Reply:
@dlipicar right, we should not add it there, cause we will lose data for step 2.